File: 0445-add-two-numbers-ii/0445-add-two-numbers-ii.kt
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
  import scala.collection.mutable
  import scala.math.Integral.Implicits._
  def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
    val s1 = new mutable.Stack[Int]()
    val s2 = new mutable.Stack[Int]()
    def reverseList(l1: ListNode, s1: mutable.Stack[Int]): mutable.Stack[Int] = l1 match {
      case null => s1
      case l1 => { s1.push(l1.x)
                   reverseList(l1.next, s1)}
    }
    reverseList(l1, s1)
    reverseList(l2, s2)

    def add(s1: mutable.Stack[Int], s2: mutable.Stack[Int], carry: Int, res: ListNode): ListNode = {
      if (s1.isEmpty && s2.isEmpty && carry == 0) return res
      val n1 = if (s1.nonEmpty) s1.pop() else 0
      val n2 = if (s2.nonEmpty) s2.pop() else 0
      val (q, r) = (carry + n1 + n2) /% 10
      if (res == null)  add(s1, s2, q, new ListNode(r))
      else add(s1, s2, q, new ListNode(r, res))
    }
    add(s1, s2, 0, null)
  }
}

File: 0445-add-two-numbers-ii/0445-add-two-numbers-ii.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        int l1Length = 0,
            l2Length = 0;
        
        ListNode tempNode = null,
                 templ1Head = new ListNode(-1),
                 templ2Head = new ListNode(-1);
        
        templ1Head.next = l1;
        templ2Head.next = l2;
        
        tempNode = l1;
        while (tempNode != null)
        {
            l1Length++;
            tempNode = tempNode.next;
        }
        
        tempNode = l2;
        while (tempNode != null)
        {
            l2Length++;
            tempNode = tempNode.next;
        }
        
        while (l1Length > l2Length)
        {
            tempNode = templ2Head.next;
            templ2Head.next = new ListNode(0);
            templ2Head.next.next = tempNode;
            l1Length--;
        }
        
        while (l1Length < l2Length)
        {
            tempNode = templ1Head.next;
            templ1Head.next = new ListNode(0);
            templ1Head.next.next = tempNode;
            l2Length--;
        }
        
        if (AddSingleNumber(templ1Head.next, templ2Head.next) == 1)
        {
            templ1Head.val = 1;
            return templ1Head;
        }
        else
            return templ1Head.next;
    }
    
    private int AddSingleNumber(ListNode node1, ListNode node2)
    {
        int temp = 0;
        
        if (node1.next == null && node2.next == null)
            temp = node1.val + node2.val;
        else
            temp = node1.val + node2.val + AddSingleNumber(node1.next, node2.next);
            
        if (temp >= 10)
        {
            node1.val = temp % 10;
            return 1;
        }
        else
        {
            node1.val = temp;
            return 0;
        }
    }
}